### PR TITLE
config: Add --whitelist support.

### PR DESCRIPTION
**Pull request by dajohi**
--whitelist specifies IP networks and IPs of remote peers that will not
be banned for misbehaving.

### CODE REVIEW

**Comment by stevenroose** [comment]
This would fix #697.

File: server.go [hunk 1]
```diff
@@ -1569,11 +1575,41 @@ func (s *server) listenHandler(listener net.Listener) {
...
+			sp.isWhitelisted = s.isWhitelisted(addr)
+		}
 	}
 	s.wg.Done()
 	srvrLog.Tracef("Listener handler done for %s", listener.Addr())
 }
 
+// isWhitelisted returns whether the IP address is included in the whitelisted
+// networks and IPs.
+func (s *server) isWhitelisted(addr net.Addr) bool {
```
Comment (davecgh) [inline server.go]: Any reason this is defined on the server type as opposed to being a standalone function?  The code doesn't appear to use anything from the server instance.

**Comment by davecgh** [comment]
I'm fairly certain this will not work well with tor because `.RemoteAddr()` on a proxy connection is the proxy itself.  Thus, if you were to do the most common proxy setup of using tor (which will be on 127.0.0.1) and whitelist 127.0.0.1, you'd effectively be whitelisting all nodes as this is currently written.

**Comment by dajohi** [comment]
The user should run Tor on a different IP then.

**Comment by davecgh** [comment]
This is a NACK until the localhost issue via proxy can be resolved.

**Comment by davecgh** [comment]
After thinking about this some more, I think there should be check to see if both a proxy address and a whitelist addresses overlap and, if they do, disallow it.  That would solve the above concerns.

**Comment by dajohi** [comment]
rebased.

**Comment by dajohi** [comment]
Rebased

**Pull request merged**